doc/neps/nep-0049.rst

```diff
@@ -0,0 +1,338 @@
+===================================
+NEP 49 — Data allocation strategies
+===================================
+
+:Author: Matti Picus
+:Status: Draft
+:Type: Standards Track
+:Created: 2021-04-18
+
+
+Abstract
+--------
+
+The ``numpy.ndarray`` requires additional memory allocations
+to hold ``numpy.ndarray.strides``, ``numpy.ndarray.shape`` and
+``numpy.ndarray.data`` attributes. These attributes are specially allocated
+after creating the python object in ``__new__`` method. The ``strides`` and
+``shape`` are stored in a piece of memory allocated internally.
+
+These allocations are small relative to the ``data``, the homogeneous chunk of
+memory used to store the actual array values (which could be pointers in the
+case of ``object`` arrays). This NEP proposes a mechanism to override the
+default memory management strategy in NumPy with user-provided alternatives.
+
+Motivation and Scope
+--------------------
+
+Users may wish to override the internal data memory routines with ones of their
+own. Two such use-cases are to ensure data alignment and to pin certain
+allocations to certain NUMA cores. The NEP is based on `PR 17582`_ and
+largely taken from the documentation added there.
+
+User who wish to change the NumPy data memory management routines will use
+:c:func:`PyDataMem_SetHandler`, which uses a :c:type:`PyDataMem_Handler`
+structure to hold pointers to functions used to manage the data memory. The
+calls are wrapped by internal routines to call :c:func:`PyTraceMalloc_Track`,
+:c:func:`PyTraceMalloc_Untrack`, and will use the
+:c:func:`PyDataMem_EventHookFunc` mechanism  already present in NumPy for
+auditing purposes. 
+
+Since a call to ``PyDataMem_SetHandler`` will change the default functions, but
+that function may be called during the lifetime of an ``ndarray`` object, each
+``ndarray`` will carry with it the ``PyDataMem_Handler`` struct used at the
+time of its instantiation, and these will be used to reallocate or free the
+data memory of the instance.
+
+Usage and Impact
+----------------
+
+The new functions can only be accessed via the NumPy C-API. An example is
+included later in the NEP. The added ``struct`` will increase the size of the
+``ndarray`` object. It is one of the major drawbacks of this approach. We can
+be reasonably sure that the change in size will have a minimal impact on
+end-user code because NumPy version 1.20 already changed the object size.
+
+Backward compatibility
+----------------------
+
+The design will not break backward compatibility. Projects that were assigning
+to the ``ndarray->data`` pointer were already breaking the current memory
+management strategy (backed by ``npy_alloc_cache``) and should restore
+``ndarray->data`` before calling ``Py_DECREF``. As mentioned above, the change
+in size should not impact end-users.
```

Comment:
This is not quite true. The old strategy was backward compatible. The worst thing that could happen is that you allocate a huge chunk, create an array with it, set OWNDATA and then the free will put it into a "too large" bucket (wasting memory).  But overall, there was never any API change the way the cache was implemented.

This PR has a tiny chance of that, because _if_ an external project would replace `ndarray->data` (i.e. realloc or use free/malloc), that would have worked before but it will stop working if the allocation strategy was modified.

Now, I am not worried about that. Nobody should be doing that, and it can only be a problem if you use one project modifying the allocation strategy AND a project modifying `ndarray->data` together.

Your PR makes sure that if the user provides the data (and may have allocated it), we will end up using the normal `free` strategy, so no issue exists for arrays created from passed in `data`.

Reply:
(same also applies to `->shape`, but that is even crazier, since the shape and strides were always allocated together, so you have to dig into NumPy code to get it right.  Overall, it makes me almost wonder if we could get away with breaking ABI and replacing the allocation strategy there with `PyMem_Malloc()` by default, but I guess it doesn't matter.)

Reply:
I didn't change the shape/strides allocation in the PR. Should I make it explicit in the NEP that the policy only touches the data and not the shape/strides ?

Reply:
Yeah, I think I misread the abstract a bit and got sidetracked.  Maybe rearrange the second paragraph of the abstract slightly would set the track clearer (brainstorming here):

> This NEP proposes a mechanism to override the memory management strategy used for `ndarray->data` with user-provided alternatives.  This allocation holds the arrays data and is can be very large.  As this is the data that operations work on it can benefit from custom allocation strategies to guarantee data alignment or placement in memory shared with the GPU.

Reply:
Updated.